src/query/query_hash_scan.c

```diff
@@ -0,0 +1,421 @@
+    {
+      if (key->free_values)
+	{
+	  for (i = 0; i < key->val_count; i++)
+	    {
+	      if (key->values[i])
+		{
+		  pr_free_value (key->values[i]);
+		}
+	    }
+	}
+
+      /* free values array */
+      db_private_free (thread_p, key->values);
+    }
+
+  /* free structure */
+  db_private_free (thread_p, key);
+}
+
+/*
+ * qdata_hash_scan_key () - compute hash of aggregate key
+ *   returns: hash value
+ *   key(in): key
+ *   ht_size(in): hash table size (in buckets)
+ */
+unsigned int
+qdata_hash_scan_key (const void *key, unsigned int ht_size)
+{
+  HASH_SCAN_KEY *ckey = (HASH_SCAN_KEY *) key;
+  unsigned int hash_val = 0, tmp_hash_val;
+  int i;
+
+  /* build hash value */
+  for (i = 0; i < ckey->val_count; i++)
+    {
+      tmp_hash_val = mht_get_hash_number (ht_size, ckey->values[i]);
+      hash_val = hash_val ^ tmp_hash_val;
+      if (hash_val == 0)
+	{
+	  hash_val = tmp_hash_val;
+	}
+    }
+
+  return hash_val;
+}
+
+/*
+ * qdata_hscan_key_compare () - compare two aggregate keys
+ *   returns: comparison result
+ *   key1(in): first key
+ *   key2(in): second key
+ *   diff_pos(out): if not equal, position of difference, otherwise -1
+ */
+static DB_VALUE_COMPARE_RESULT
+qdata_hscan_key_compare (HASH_SCAN_KEY * ckey1, HASH_SCAN_KEY * ckey2, int *diff_pos)
+{
+  DB_VALUE_COMPARE_RESULT result;
+  int i;
+
+  assert (diff_pos);
+  *diff_pos = -1;
+
+  if (ckey1 == ckey2)
+    {
+      /* same pointer, same values */
+      return DB_EQ;
+    }
+
+  if (ckey1->val_count != ckey2->val_count)
+    {
+      /* can't compare keys of different sizes; shouldn't get here */
+      assert (false);
+      return DB_UNK;
+    }
+
+  for (i = 0; i < ckey1->val_count; i++)
+    {
+      result = tp_value_compare (ckey1->values[i], ckey2->values[i], 0, 1);
+      if (result != DB_EQ)
+	{
+	  *diff_pos = i;
+	  return result;
+	}
+    }
+
+  /* if we got this far, it's equal */
+  return DB_EQ;
+}
+
+/*
+ * qdata_hscan_key_eq () - check equality of two aggregate keys
+ *   returns: true if equal, false otherwise
+ *   key1(in): first key
+ *   key2(in): second key
+ */
+int
+qdata_hscan_key_eq (const void *key1, const void *key2)
+{
+  int decoy;
+
+  /* compare for equality */
+  return (qdata_hscan_key_compare ((HASH_SCAN_KEY *) key1, (HASH_SCAN_KEY *) key2, &decoy) == DB_EQ);
+}
+
+/*
+ * qdata_build_hscan_key () - build aggregate key structure from reguvar list
+ *   returns: NO_ERROR or error code
+ *   thread_p(in): thread
+ *   key(out): aggregate key
+ *   regu_list(in): reguvar list for fetching values
+ */
+int
+qdata_build_hscan_key (THREAD_ENTRY * thread_p, val_descr * vd, REGU_VARIABLE_LIST regu_list, HASH_SCAN_KEY * key)
+{
+  int rc = NO_ERROR;
+
+  /* build key */
+  key->free_values = false;	/* references precreated DB_VALUES */
+  key->val_count = 0;
+  while (regu_list != NULL)
+    {
+      rc = fetch_peek_dbval (thread_p, &regu_list->value, vd, NULL, NULL, NULL, &key->values[key->val_count]);
```

Comment:
don't need to check key->values[key->val_count] is out of bound?

Reply:
There is no problem because it allocates memory as much as key->val_count.
qdata_alloc_hscan_key ()
```
  key->values = (DB_VALUE **) db_private_alloc (thread_p, sizeof (DB_VALUE *) * val_cnt);
...
  key->val_count = val_cnt;
```